analyses/metacells/scripts/run-seacells.py

```diff
@@ -0,0 +1,159 @@
+#!/usr/bin/env python3
+
+"""
+Seacells Analysis Script
+Joshua Shapiro
+2024-07-18
+
+This script runs the SEACells algorithm on a dataset.
+"""
+
+# Load modules
+#
+# Load required Python modules at the top of your script
+# We have included the standard `pathlib` module and the `session_info` module
+# that we will be using at the bottom of this notebook to record the versions of
+# the modules used in this analysis.
+#
+# Do not install modules here; only load them with `import` statements.
+# Avoid renaming modules with `as` statements, unless you are performing a
+# standard renaming (e.g., `import pandas as pd`).
+
+import argparse
+import contextlib
+import pathlib
+import pickle
+import sys
+
+import anndata
+import numpy as np
+import scanpy as sc
+import SEACells
+import session_info
+
+
+def convert_adata(adata: anndata.AnnData) -> anndata.AnnData:
+    """
+    Convert an ScPCA AnnData object to the formatting expected by scanpy and SEACells.
+    """
+    # put the highly variable genes in a column of adata.var
+    adata.var["highly_variable"] = adata.var.gene_ids.isin(
+        adata.uns["highly_variable_genes"]
+    )
+
+    # recompute principal components and umap (adds additional metadata)
+    sc.tl.pca(adata, n_comps=50, mask_var="highly_variable")
+    sc.pp.neighbors(adata)
+    sc.tl.umap(adata)
+
+    return adata
+
+
+def run_seacells(
+    adata: anndata.AnnData, cell_ratio: float = 75, verbose: bool = False
+) -> tuple[anndata.AnnData, SEACells.core.SEACells]:
+    """
+    Run the SEACells algorithm on the given dataset.
+
+    Parameters
+    ----------
+    adata : anndata.AnnData
+        An annData object containing the data to run the SEACells algorithm on.
+        Should contain an X_pca field with the PCA coordinates of the cells.
+    cell_ratio : float
+        The ratio of cells to metacells to use; i.e. number of cells per metacell
+    verbose : bool
+        Whether to print verbose output during the SEACells algorithm
+
+    Returns
+    -------
+    anndata.AnnData
+        The input AnnData object with the metacell assignments added to the obs table with the key "SEACell"
+    SEACells.core.SEACells
+        The SEACells model object
+    """
+    n_metacells = round(adata.n_obs / cell_ratio)
+    n_eigs = 10  # number of eigenvalues for initialization
+    # initialize the SEACells model
+    model = SEACells.core.SEACells(
+        adata,
+        build_kernel_on="X_pca",
+        n_SEACells=n_metacells,
+        n_waypoint_eigs=n_eigs,
+        convergence_epsilon=1e-5,
+        verbose=verbose,
+    )
+    # initialize and fit model
+    model.construct_kernel_matrix()
+    model.initialize_archetypes()
+    model.fit(min_iter=10, max_iter=50)
```

Comment:
I think this is the function - any reason to not keep their `max_iter=100`? Is it super slow?
https://github.com/dpeerlab/SEACells/blob/3462c624ffae0df6d3930490f345f00196c3503e/SEACells/cpu.py#L653-L659